### PR TITLE
[jit] basic testing of builtin alias annotations

### PR DESCRIPTION
**Pull request by suo**
Check whether the codegen'd alias annotations actually track alias creation and writes correctly. This could be made more exhaustive, but it's good enough for now.

### CODE REVIEW

File: torch/csrc/jit/constants.cpp [hunk 1]
```diff
@@ -129,8 +130,47 @@ RegisterOperators reg({
 });
 
 c10::optional<IValue> toIValue(const Value* v) {
-  if(v->node()->kind() != prim::Constant)
+  if (v->node()->kind() == prim::ListConstruct) {
```
Comment (zdevito) [inline torch/csrc/jit/constants.cpp]: I don't like this additional logic here. It is essentially duplicating constant propagation.  If this is just needed to make the test harness work, can it just be put in that source file as a separate function? Everything that is constant should eventually become a prim::Constant in the

**Review by zdevito — APPROVED** [review]
Looks good! See one comment about preventing the duplication of constant propagation logic.

File: torch/csrc/jit/constants.cpp (hunk 1 above)
Comment (suo) [reply torch/csrc/jit/constants.cpp]: sure, that makes sense

File: aten/src/ATen/core/ivalue.h [hunk 2]
```diff
@@ -132,11 +132,36 @@ struct CAFFE2_API IValue final {
...
+
+    // Tensors should be compared based on internal storage
+    if (this->isTensor()) {
+      const auto thisTensor = this->toTensor();
+      const auto rhsTensor = rhs.toTensor();
+      return thisTensor.is_alias_of(rhsTensor);
+    }
+
+    // Other types can be compared by their ptr value
+    return this->payload.as_intrusive_ptr == rhs.payload.as_intrusive_ptr;
```
Comment (apaszke) [inline aten/src/ATen/core/ivalue.h]: Ugh that's a pretty strong assumption about aliasing of other types. We should probably rename this to `mayAlias` and return `true` in this case.

File: torch/csrc/jit/passes/utils/check_alias_annotation.cpp [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    ret.push_back(deepCopy(v));
+  }
+  return ret;
+}
+
+bool deepEquals(const IValue& lhs, const IValue& rhs) {
+  // only check tensors for now
+  if (!lhs.isTensor() || !rhs.isTensor()) {
+    return true;
+  }
```
Comment (apaszke) [inline torch/csrc/jit/passes/utils/check_alias_annotation.cpp]: Oh god that's so misleading. I'd assume `deepEquals(1, 2)` to be false (similarly `deepEquals(torch.tensor(1.), 2)`) but the return true! We should throw in that case!

Comment (suo) [reply torch/csrc/jit/passes/utils/check_alias_annotation.cpp]: yeah this is terrible. I'll put up a PR to clean up this testing stuff.

**Pull request closed without merging**